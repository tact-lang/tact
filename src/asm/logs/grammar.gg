vmLine =
      VmLoc
    / VmStack
    / VmExecute
    / VmLimitChanged
    / VmGasRemaining
    / VmException
    / VmExceptionHandler
    / VmFinalC5
    / VmUnknown;

VmStack            = "stack: " stack:$[^\n]+;
VmLoc              = "code cell hash:" hash:hex "offset:" offset:number;
VmExecute          = "execute " instr:$[^\n]+;
VmLimitChanged     = "changing gas limit to " limit:number;
VmGasRemaining     = "gas remaining: " gas:number;
VmException        = "handling exception code " errno:number ": " message:$[^\n]*;
VmExceptionHandler = "default exception handler, terminating vm with exit code" errno:number;
VmFinalC5          = "final c5:" value:Cell;
VmUnknown          = !"stack" text:$[^\n]+ "\n";

VmParsedStack = "[" values:VmStackValue* "]";

VmStackValue = value:(
      Null
    / NaN
    / Integer
    / Tuple
    / TupleParen
    / Cell
    / Continuation
    / Builder
    / CellSlice
    / Unknown
);

Null         = "()" / "(null)";
NaN          = "NaN";
Integer      = value:number;
Tuple        = "[" elements:VmStackValue* "]";
TupleParen   = "(" elements:VmStackValue* ")";
Cell         = "C{" value:hex "}";
Continuation = "Cont{" value:$#([A-Za-z_0-9]*) "}";
Builder      = "BC{" value:hex "}";
Unknown      = "???";

CellSlice     = "CS{" body:(CellSliceBody / CellSliceShortBody) "}";
CellSliceBody = "Cell{" value:hex "}" bits:CellSliceBits ";" refs:CellSliceRefs;
CellSliceBits = "bits:" start:number ".." end:number;
CellSliceRefs = "refs:" start:number ".." end:number;

CellSliceShortBody = value:hex;

number   = op:"-"? value:#$(digit+);
digit    = [0-9];

hexDigit "hexadecimal digit" = [0-9a-fA-F];
hex      = $#(value:hexDigit*);

space "space" = [ \t\r\n];
