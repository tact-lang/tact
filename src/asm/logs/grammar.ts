/* Generated. Do not edit. */
/* eslint-disable @typescript-eslint/no-namespace */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-redundant-type-constituents */
/* eslint-disable @typescript-eslint/no-duplicate-type-constituents */
/* eslint-disable @typescript-eslint/no-unused-vars */
import * as $ from "@tonstudio/parser-runtime"
export namespace $ast {
    export type VmLoc = $.Located<{
        readonly $: "VmLoc"
        readonly hash: hex
        readonly offset: $number
    }>
    export type VmStack = $.Located<{
        readonly $: "VmStack"
        readonly stack: string
    }>
    export type VmExecute = $.Located<{
        readonly $: "VmExecute"
        readonly instr: string
    }>
    export type VmLimitChanged = $.Located<{
        readonly $: "VmLimitChanged"
        readonly limit: $number
    }>
    export type VmGasRemaining = $.Located<{
        readonly $: "VmGasRemaining"
        readonly gas: $number
    }>
    export type VmException = $.Located<{
        readonly $: "VmException"
        readonly errno: $number
        readonly message: string
    }>
    export type VmExceptionHandler = $.Located<{
        readonly $: "VmExceptionHandler"
        readonly errno: $number
    }>
    export type VmFinalC5 = $.Located<{
        readonly $: "VmFinalC5"
        readonly value: Cell
    }>
    export type VmUnknown = $.Located<{
        readonly $: "VmUnknown"
        readonly text: string
    }>
    export type vmLine =
        | VmLoc
        | VmStack
        | VmExecute
        | VmLimitChanged
        | VmGasRemaining
        | VmException
        | VmExceptionHandler
        | VmFinalC5
        | VmUnknown
    export type VmParsedStack = $.Located<{
        readonly $: "VmParsedStack"
        readonly values: readonly VmStackValue[]
    }>
    export type VmStackValue = $.Located<{
        readonly $: "VmStackValue"
        readonly value:
            | Null
            | NaN
            | Integer
            | Tuple
            | TupleParen
            | Cell
            | Continuation
            | Builder
            | CellSlice
            | Unknown
    }>
    export type Null = $.Located<{
        readonly $: "Null"
    }>
    export type NaN = $.Located<{
        readonly $: "NaN"
    }>
    export type Integer = $.Located<{
        readonly $: "Integer"
        readonly value: $number
    }>
    export type Tuple = $.Located<{
        readonly $: "Tuple"
        readonly elements: readonly VmStackValue[]
    }>
    export type TupleParen = $.Located<{
        readonly $: "TupleParen"
        readonly elements: readonly VmStackValue[]
    }>
    export type Cell = $.Located<{
        readonly $: "Cell"
        readonly value: hex
    }>
    export type Continuation = $.Located<{
        readonly $: "Continuation"
        readonly value: string
    }>
    export type Builder = $.Located<{
        readonly $: "Builder"
        readonly value: hex
    }>
    export type Unknown = $.Located<{
        readonly $: "Unknown"
    }>
    export type CellSlice = $.Located<{
        readonly $: "CellSlice"
        readonly body: CellSliceBody | CellSliceShortBody
    }>
    export type CellSliceBody = $.Located<{
        readonly $: "CellSliceBody"
        readonly value: hex
        readonly bits: CellSliceBits
        readonly refs: CellSliceRefs
    }>
    export type CellSliceBits = $.Located<{
        readonly $: "CellSliceBits"
        readonly start: $number
        readonly end: $number
    }>
    export type CellSliceRefs = $.Located<{
        readonly $: "CellSliceRefs"
        readonly start: $number
        readonly end: $number
    }>
    export type CellSliceShortBody = $.Located<{
        readonly $: "CellSliceShortBody"
        readonly value: hex
    }>
    export type $number = {
        readonly op: "-" | undefined
        readonly value: string
    }
    export type digit = string
    export type hexDigit = string | string | string
    export type hex = string
    export type space = " " | "\t" | "\r" | "\n"
}
export const VmLoc: $.Parser<$ast.VmLoc> = $.loc(
    $.field(
        $.pure("VmLoc"),
        "$",
        $.right(
            $.str("code cell hash:"),
            $.field(
                $.lazy(() => hex),
                "hash",
                $.right(
                    $.str("offset:"),
                    $.field(
                        $.lazy(() => $number),
                        "offset",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const VmStack: $.Parser<$ast.VmStack> = $.loc(
    $.field(
        $.pure("VmStack"),
        "$",
        $.right(
            $.str("stack: "),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "stack",
                $.eps,
            ),
        ),
    ),
)
export const VmExecute: $.Parser<$ast.VmExecute> = $.loc(
    $.field(
        $.pure("VmExecute"),
        "$",
        $.right(
            $.str("execute "),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "instr",
                $.eps,
            ),
        ),
    ),
)
export const VmLimitChanged: $.Parser<$ast.VmLimitChanged> = $.loc(
    $.field(
        $.pure("VmLimitChanged"),
        "$",
        $.right(
            $.str("changing gas limit to "),
            $.field(
                $.lazy(() => $number),
                "limit",
                $.eps,
            ),
        ),
    ),
)
export const VmGasRemaining: $.Parser<$ast.VmGasRemaining> = $.loc(
    $.field(
        $.pure("VmGasRemaining"),
        "$",
        $.right(
            $.str("gas remaining: "),
            $.field(
                $.lazy(() => $number),
                "gas",
                $.eps,
            ),
        ),
    ),
)
export const VmException: $.Parser<$ast.VmException> = $.loc(
    $.field(
        $.pure("VmException"),
        "$",
        $.right(
            $.str("handling exception code "),
            $.field(
                $.lazy(() => $number),
                "errno",
                $.right(
                    $.str(": "),
                    $.field(
                        $.stry($.star($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                        "message",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const VmExceptionHandler: $.Parser<$ast.VmExceptionHandler> = $.loc(
    $.field(
        $.pure("VmExceptionHandler"),
        "$",
        $.right(
            $.str("default exception handler, terminating vm with exit code"),
            $.field(
                $.lazy(() => $number),
                "errno",
                $.eps,
            ),
        ),
    ),
)
export const VmFinalC5: $.Parser<$ast.VmFinalC5> = $.loc(
    $.field(
        $.pure("VmFinalC5"),
        "$",
        $.right(
            $.str("final c5:"),
            $.field(
                $.lazy(() => Cell),
                "value",
                $.eps,
            ),
        ),
    ),
)
export const VmUnknown: $.Parser<$ast.VmUnknown> = $.loc(
    $.field(
        $.pure("VmUnknown"),
        "$",
        $.right(
            $.lookNeg($.str("stack")),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "text",
                $.right($.str("\n"), $.eps),
            ),
        ),
    ),
)
export const vmLine: $.Parser<$ast.vmLine> = $.alt(
    VmLoc,
    $.alt(
        VmStack,
        $.alt(
            VmExecute,
            $.alt(
                VmLimitChanged,
                $.alt(
                    VmGasRemaining,
                    $.alt(VmException, $.alt(VmExceptionHandler, $.alt(VmFinalC5, VmUnknown))),
                ),
            ),
        ),
    ),
)
export const VmParsedStack: $.Parser<$ast.VmParsedStack> = $.loc(
    $.field(
        $.pure("VmParsedStack"),
        "$",
        $.right(
            $.str("["),
            $.field($.star($.lazy(() => VmStackValue)), "values", $.right($.str("]"), $.eps)),
        ),
    ),
)
export const VmStackValue: $.Parser<$ast.VmStackValue> = $.loc(
    $.field(
        $.pure("VmStackValue"),
        "$",
        $.field(
            $.alt(
                $.lazy(() => Null),
                $.alt(
                    $.lazy(() => NaN),
                    $.alt(
                        $.lazy(() => Integer),
                        $.alt(
                            $.lazy(() => Tuple),
                            $.alt(
                                $.lazy(() => TupleParen),
                                $.alt(
                                    $.lazy(() => Cell),
                                    $.alt(
                                        $.lazy(() => Continuation),
                                        $.alt(
                                            $.lazy(() => Builder),
                                            $.alt(
                                                $.lazy(() => CellSlice),
                                                $.lazy(() => Unknown),
                                            ),
                                        ),
                                    ),
                                ),
                            ),
                        ),
                    ),
                ),
            ),
            "value",
            $.eps,
        ),
    ),
)
export const Null: $.Parser<$ast.Null> = $.loc(
    $.field($.pure("Null"), "$", $.right($.alt($.str("()"), $.str("(null)")), $.eps)),
)
export const NaN: $.Parser<$ast.NaN> = $.loc(
    $.field($.pure("NaN"), "$", $.right($.str("NaN"), $.eps)),
)
export const Integer: $.Parser<$ast.Integer> = $.loc(
    $.field(
        $.pure("Integer"),
        "$",
        $.field(
            $.lazy(() => $number),
            "value",
            $.eps,
        ),
    ),
)
export const Tuple: $.Parser<$ast.Tuple> = $.loc(
    $.field(
        $.pure("Tuple"),
        "$",
        $.right($.str("["), $.field($.star(VmStackValue), "elements", $.right($.str("]"), $.eps))),
    ),
)
export const TupleParen: $.Parser<$ast.TupleParen> = $.loc(
    $.field(
        $.pure("TupleParen"),
        "$",
        $.right($.str("("), $.field($.star(VmStackValue), "elements", $.right($.str(")"), $.eps))),
    ),
)
export const Cell: $.Parser<$ast.Cell> = $.loc(
    $.field(
        $.pure("Cell"),
        "$",
        $.right(
            $.str("C{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Continuation: $.Parser<$ast.Continuation> = $.loc(
    $.field(
        $.pure("Continuation"),
        "$",
        $.right(
            $.str("Cont{"),
            $.field(
                $.stry(
                    $.lex(
                        $.star(
                            $.regex<string | string | "_" | string>("A-Za-z_0-9", [
                                $.ExpRange("A", "Z"),
                                $.ExpRange("a", "z"),
                                $.ExpString("_"),
                                $.ExpRange("0", "9"),
                            ]),
                        ),
                    ),
                ),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Builder: $.Parser<$ast.Builder> = $.loc(
    $.field(
        $.pure("Builder"),
        "$",
        $.right(
            $.str("BC{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Unknown: $.Parser<$ast.Unknown> = $.loc(
    $.field($.pure("Unknown"), "$", $.right($.str("???"), $.eps)),
)
export const CellSlice: $.Parser<$ast.CellSlice> = $.loc(
    $.field(
        $.pure("CellSlice"),
        "$",
        $.right(
            $.str("CS{"),
            $.field(
                $.alt(
                    $.lazy(() => CellSliceBody),
                    $.lazy(() => CellSliceShortBody),
                ),
                "body",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const CellSliceBody: $.Parser<$ast.CellSliceBody> = $.loc(
    $.field(
        $.pure("CellSliceBody"),
        "$",
        $.right(
            $.str("Cell{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right(
                    $.str("}"),
                    $.field(
                        $.lazy(() => CellSliceBits),
                        "bits",
                        $.right(
                            $.str(";"),
                            $.field(
                                $.lazy(() => CellSliceRefs),
                                "refs",
                                $.eps,
                            ),
                        ),
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceBits: $.Parser<$ast.CellSliceBits> = $.loc(
    $.field(
        $.pure("CellSliceBits"),
        "$",
        $.right(
            $.str("bits:"),
            $.field(
                $.lazy(() => $number),
                "start",
                $.right(
                    $.str(".."),
                    $.field(
                        $.lazy(() => $number),
                        "end",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceRefs: $.Parser<$ast.CellSliceRefs> = $.loc(
    $.field(
        $.pure("CellSliceRefs"),
        "$",
        $.right(
            $.str("refs:"),
            $.field(
                $.lazy(() => $number),
                "start",
                $.right(
                    $.str(".."),
                    $.field(
                        $.lazy(() => $number),
                        "end",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceShortBody: $.Parser<$ast.CellSliceShortBody> = $.loc(
    $.field(
        $.pure("CellSliceShortBody"),
        "$",
        $.field(
            $.lazy(() => hex),
            "value",
            $.eps,
        ),
    ),
)
export const $number: $.Parser<$ast.$number> = $.field(
    $.opt($.str("-")),
    "op",
    $.field($.lex($.stry($.plus($.lazy(() => digit)))), "value", $.eps),
)
export const digit: $.Parser<$ast.digit> = $.regex<string>("0-9", [$.ExpRange("0", "9")])
export const hexDigit: $.Parser<$ast.hexDigit> = $.named(
    "hexadecimal digit",
    $.regex<string | string | string>("0-9a-fA-F", [
        $.ExpRange("0", "9"),
        $.ExpRange("a", "f"),
        $.ExpRange("A", "F"),
    ]),
)
export const hex: $.Parser<$ast.hex> = $.stry($.lex($.field($.star(hexDigit), "value", $.eps)))
export const space: $.Parser<$ast.space> = $.named(
    "space",
    $.regex<" " | "\t" | "\r" | "\n">(" \\t\\r\\n", [
        $.ExpString(" "),
        $.ExpString("\t"),
        $.ExpString("\r"),
        $.ExpString("\n"),
    ]),
)
