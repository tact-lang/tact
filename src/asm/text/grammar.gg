File = instructions:instructions;

instructions = (Instruction / ExplicitRef / EmbedSlice / Exotic)*;

Instruction = name:Id args:Argument*;

Argument = expression:(
      IntegerLiteral
    / DataLiteral
    / Code
    / Dictionary
    / StackElement
    / ControlRegister
);

StackElement = value:$#("s" "-"? IntegerLiteralDec);
ControlRegister = value:$#("c" IntegerLiteralDec);

Code = "{" instructions:instructions "}";

Dictionary = "[" entries:DictionaryEntry* "]";

DictionaryEntry = id:IntegerLiteral "=>" code:Code;

ExplicitRef = keyword<"ref"> code:Code;
EmbedSlice  = keyword<"embed"> data:DataLiteral;
Exotic      = keyword<"exotic"> lib:(ExoticLibrary / DefaultExotic);

ExoticLibrary = keyword<"library"> data:DataLiteral;
DefaultExotic = data:DataLiteral;

idPart "identifier character" = [a-zA-Z0-9_];
Id "identifier" = name:#$(!reservedWord [a-zA-Z_] idPart*);

DataLiteral = value:(HexLiteral / BinLiteral / BocLiteral / StringLiteral);

HexLiteral = "x{" content:$(hexDigit* "_"?) "}";
BinLiteral = "b{" content:$(binDigit*) "}";
BocLiteral = "boc{" content:$(hexDigit*) "}";

// Literals
IntegerLiteral = op:"-"? value:(IntegerLiteralHex / IntegerLiteralBin / IntegerLiteralOct / IntegerLiteralDec);

IntegerLiteralDec = digits:#underscored<digit>;
IntegerLiteralHex = digits:#("0" [xX] @underscored<hexDigit>);
IntegerLiteralBin = digits:#("0" [bB] @underscored<binDigit>);
IntegerLiteralOct = digits:#("0" [oO] @underscored<[0-7]>);

underscored<T> = $(T ("_"? T)*);

hexDigit "hexadecimal digit" = [0-9a-fA-F];
binDigit "binary digit" = [01];
digit "digit" = [0-9];

StringLiteral = value:#("\"" @$([^"\\])* "\"");

keyword<T> = #(@T !idPart);

reservedWord "reserved word" = keyword<(
    "ref" / "embed" / "exotic" / "library"
)>;

space "space" = [ \t\r\n] / comment;
comment = singleLineComment;
singleLineComment = "//" @$[^\r\n]*;
