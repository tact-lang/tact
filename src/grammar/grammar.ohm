Tact {

    // Starting point of the program
    Program = ProgramItem*
    ProgramItem = Struct
                | Contract
                | Primitive
                | StaticFunction
                | NativeFunction
                | ProgramImport
                | Trait
                | Constant
    ProgramImport = import stringLiteral ";"

    // Built-in declarations
    Primitive = "primitive" Type ";"

    // Static function
    StaticFunction = Function
    NativeFunction = nameAttribute "(" funcId ")" FunctionAttribute* native id "(" ListOf<FunctionArg,","> ","? ")" ";" --withVoid
                   | nameAttribute "(" funcId ")" FunctionAttribute* native id "(" ListOf<FunctionArg,","> ","? ")" ":" Type ";" --withType
    
    // Field declarations
    Type = typeLiteral "?" --optional
         | typeLiteral --required
         | "map" "<" typeLiteral (as id)? "," typeLiteral (as id)? ">" --map
         | "bounced" "<" typeLiteral ">" --bounced
    Field = id ":" Type ";" --default
          | id ":" Type "=" Expression ";" --defaultWithInit
          | id ":" Type as id ";" --withSerialization
          | id ":" Type as id "=" Expression ";" --withSerializationAndInit
    
    // Constant
    ConstantAttribute = virtual    --virtual
                      | override   --override
                      | abstract   --abstract
    Constant = ConstantAttribute* ~fun const id ":" Type "=" Expression ";" --withValue
             | ConstantAttribute* ~fun const id ":" Type ";"                --withEmpty

    // Struct
    Struct = "struct" typeLiteral "{" StructBody* "}" --originary
           | "message" typeLiteral "{" StructBody* "}" --message
           | "message" "(" integerLiteral ")" typeLiteral "{" StructBody* "}" --messageWithId
    StructBody = Field

    // Contract
    Contract = ContractAttribute* contract id "{" ContractBody* "}" --simple
             | ContractAttribute* contract id with NonemptyListOf<id,","> ","? "{" ContractBody* "}" --withTraits
    ContractInit = "init" "(" ListOf<FunctionArg,","> ","? ")" "{" Statement* "}"
    ContractBody = Field
                 | ContractInit
                 | ReceiveFunction
                 | Function
                 | Constant
    
    // Trait
    Trait = ContractAttribute* trait id "{" TraitBody* "}" --originary
          | ContractAttribute* trait id with NonemptyListOf<id,","> ","? "{" TraitBody* "}" --withTraits
    TraitBody = Field
              | ReceiveFunction
              | Function
              | Constant

    // Contract attributes
    ContractAttribute = "@interface" "(" stringLiteral ")" --interface

    // Function
    FunctionAttribute = "get"     --getter
                      | mutates   --mutates
                      | extends   --extends
                      | virtual   --virtual
                      | override  --override
                      | inline    --inline
                      | abstract  --abstract
    Function = FunctionAttribute* fun id "(" ListOf<FunctionArg,","> ","? ")" "{" Statement* "}" --withVoid
             | FunctionAttribute* fun id "(" ListOf<FunctionArg,","> ","? ")" ":" Type "{" Statement* "}" --withType
             | FunctionAttribute* fun id "(" ListOf<FunctionArg,","> ","? ")" ";" --abstractVoid
             | FunctionAttribute* fun id "(" ListOf<FunctionArg,","> ","? ")" ":" Type ";" --abstractType
    FunctionArg = id ":" Type
    
    ReceiveFunction = "receive" "(" FunctionArg ")" "{" Statement* "}" --simple
                    | "receive" "(" ")" "{" Statement* "}" --empty
                    | "receive" "(" stringLiteral ")" "{" Statement* "}" --comment
                    | "bounced" "(" FunctionArg ")" "{" Statement* "}" --bounced
                    | "external" "(" FunctionArg ")" "{" Statement* "}" --externalSimple
                    | "external" "(" stringLiteral ")" "{" Statement* "}" --externalComment
                    | "external" "(" ")" "{" Statement* "}" --externalEmpty

    // Statements
    Statement = StatementLet
              | StatementBlock
              | StatementReturn
              | StatementExpression
              | StatementAssign
              | StatementAugmentedAssign
              | StatementCondition
              | StatementWhile
              | StatementRepeat
              | StatementUntil
              | StatementTry
    StatementBlock = "{" Statement* "}"
    StatementLet = let id ":" Type "=" Expression ";"
    StatementReturn = return Expression ";" --withExpression
                    | return ";" --withoutExpression    
    StatementExpression = Expression ";"
    StatementAssign = LValue "=" Expression ";"
    StatementAugmentedAssign = StatementAugmentedAssignAdd
                             | StatementAugmentedAssignSub
                             | StatementAugmentedAssignMul
                             | StatementAugmentedAssignDiv
                             | StatementAugmentedAssignRem
    StatementAugmentedAssignAdd = LValue "+=" Expression ";"
    StatementAugmentedAssignSub = LValue "-=" Expression ";"
    StatementAugmentedAssignMul = LValue "*=" Expression ";"
    StatementAugmentedAssignDiv = LValue "/=" Expression ";"
    StatementAugmentedAssignRem = LValue "%=" Expression ";"
    StatementCondition = if Expression "{" Statement* "}" ~else --simple
                       | if Expression "{" Statement* "}" else "{" Statement* "}" --withElse
                       | if Expression "{" Statement* "}" else StatementCondition --withElseIf
    StatementWhile = while "(" Expression ")" "{" Statement* "}"
    StatementRepeat = repeat "(" Expression ")" "{" Statement* "}"
    StatementUntil = do "{" Statement* "}" until "(" Expression ")" ";"
    StatementTry = try "{" Statement* "}" ~catch --simple
                 | try "{" Statement* "}" catch "(" id ")" "{" Statement* "}" --withCatch

    // L-value
    LValue = id "." LValue --more
           | id --single

    // Expressions
    Expression = ExpressionConditional
    
    ExpressionConditional = ExpressionOr "?" ExpressionOr ":" ExpressionConditional --ternary
                          | ExpressionOr
    
    ExpressionOr = ExpressionOr "||" ExpressionAnd --or
                 | ExpressionAnd
    
    ExpressionAnd = ExpressionAnd "&&" ExpressionBinaryOr --and
                  | ExpressionBinaryOr
    
    ExpressionBinaryOr = ExpressionBinaryOr "|" ExpressionBinaryXor --bin_or
                       | ExpressionBinaryXor
    
    ExpressionBinaryXor = ExpressionBinaryXor "^" ExpressionBinaryAnd --bin_xor
                       | ExpressionBinaryAnd
    
    ExpressionBinaryAnd = ExpressionBinaryAnd "&" ExpressionEquality --bin_and
                        | ExpressionEquality
    
    ExpressionEquality = ExpressionEquality "!=" ExpressionCompare --not
                       | ExpressionEquality "==" ExpressionCompare --eq
                       | ExpressionCompare
    
    ExpressionCompare = ExpressionCompare ">" ExpressionBinaryShift --gt
                      | ExpressionCompare ">=" ExpressionBinaryShift --gte
                      | ExpressionCompare "<" ExpressionBinaryShift --lt
                      | ExpressionCompare "<=" ExpressionBinaryShift --lte
                      | ExpressionBinaryShift
    
    ExpressionBinaryShift = ExpressionBinaryShift "<<" ExpressionAdd --shl
                          | ExpressionBinaryShift ">>" ExpressionAdd --shr
                          | ExpressionAdd
    
    ExpressionAdd = ExpressionAdd "+" ~"+" ExpressionMul --add
                  | ExpressionAdd "-" ~"-" ExpressionMul --sub
                  | ExpressionMul
    
    ExpressionMul = ExpressionMul "*" ExpressionUnary --mul
                  | ExpressionMul "/" ExpressionUnary --div
                  | ExpressionMul "%" ExpressionUnary --rem
                  | ExpressionUnary
    
    ExpressionUnary = "-" ExpressionValue --neg
                    | "+" ExpressionValue --add
                    | "!" ExpressionValue --not
                    | ExpressionValue
    
    ExpressionBracket = "(" Expression ")"

    // Order is important
    ExpressionValue = ExpressionUnboxNotNull
                    | ExpressionCall
                    | ExpressionField
                    | ExpressionStaticCall
                    | ExpressionBracket
                    | ExpressionNew
                    | integerLiteral
                    | boolLiteral
                    | id
                    | null
                    | ExpressionInitOf
                    | ExpressionString
    ExpressionUnboxNotNull = ExpressionValue "!!"
    ExpressionString = stringLiteral
    ExpressionField = ExpressionValue "." id ~"("
    ExpressionCall = ExpressionValue "." id "(" ListOf<Expression, ","> ","? ")"
    ExpressionNew = id "{" ListOf<NewParameter, ","> ","? "}"
    NewParameter = id ":" Expression --full
                 | id                --punned 
    ExpressionStaticCall = id "(" ListOf<Expression, ","> ","? ")"
    ExpressionInitOf = initOf id "(" ListOf<Expression, ","> ","? ")"

    // Type Literal
    typeLiteral = letterAsciiUC typeLiteralPart*
    typeLiteralPart = letterAscii | digit | "_"

    // Integer Literal
    // hexDigit defined in Ohm's built-in rules (otherwise: hexDigit = "0".."9" | "a".."f" | "A".."F")
    // digit defined in Ohm's built-in rules (otherwise: digit = "0".."9")
    integerLiteral = integerLiteralHex | integerLiteralBin | integerLiteralOct | integerLiteralDec // Order is important
    integerLiteralDec = nonZeroDigit ("_"? digit)*  --nonZeroIntegerLiteralDec
                      | "0" digit*                  --integerLiteralWithLeadingZero
    integerLiteralHex = ("0x" | "0X") hexDigit ("_"? hexDigit)*
    integerLiteralBin = ("0b" | "0B") binDigit ("_"? binDigit)*
    integerLiteralOct = ("0o" | "0O") octDigit ("_"? octDigit)*
    binDigit = "0" | "1"
    octDigit = "0".."7"
    nonZeroDigit = "1".."9"

    // Letters
    letterAsciiLC = "a".."z"
    letterAsciiUC = "A".."Z"
    letterAscii = letterAsciiLC | letterAsciiUC
    letterComment = letterAsciiLC | letterAsciiUC | digit | "_"

    // ID Literal
    idStart = letterAscii | "_"
    idPart = letterAscii | digit | "_"
    id = ~reservedWord #idStart #(idPart*)

    // FunC id
    funcLetter = letterAscii | "_" | "'" | "?" | "!" | "::" | "&"
    funcId = funcLetter #(funcLetter | digit)*

    // Bool Literal
    boolLiteral = ("true" | "false") ~idPart

    // String literal
    stringLiteral = "\"" (nonQuoteOrBackslashChar | escapeSequence)* "\""
    nonQuoteOrBackslashChar = ~("\"" | "\\") any
    escapeSequence = "\\\\" -- backslash
                   | "\\\"" -- doubleQuote
                   | "\\n" -- newline
                   | "\\r" -- carriageReturn
                   | "\\t" -- tab
                   | "\\v" -- verticalTab
                   | "\\b" -- backspace
                   | "\\f" -- formFeed
                   | "\\u{" hexDigit hexDigit? hexDigit? hexDigit? hexDigit? hexDigit? "}" -- unicodeCodePoint
                   | "\\u" hexDigit hexDigit hexDigit hexDigit -- unicodeEscape
                   | "\\x" hexDigit hexDigit -- hexEscape

    // Keywords
    // NOTE Order is important
    keyword = fun 
            | let
            | return 
            | extend 
            | native 
            | public 
            | null 
            | if 
            | else 
            | while 
            | repeat 
            | do 
            | until 
            | try
            | catch
            | as  
            | mutates
            | extends
            | import
            | with
            | trait
            | initOf
            | override
            | abstract
            | virtual
            | inline
            | const
    contract = "contract" ~idPart
    let = "let" ~idPart
    fun = "fun" ~idPart
    return = "return" ~idPart
    extend = "extend" ~idPart
    native = "native" ~idPart
    public = "public" ~idPart
    null = "null" ~idPart
    if = "if" ~idPart
    else = "else" ~idPart
    while = "while" ~idPart
    repeat = "repeat" ~idPart
    do = "do" ~idPart
    until = "until" ~idPart
    try = "try" ~idPart
    catch = "catch" ~idPart
    as = "as" ~idPart
    mutates = "mutates" ~idPart
    extends = "extends" ~idPart
    import = "import" ~idPart
    with = "with" ~idPart
    trait = "trait" ~idPart
    initOf = "initOf" ~idPart
    virtual = "virtual" ~idPart
    override = "override" ~idPart
    inline = "inline" ~idPart
    const = "const" ~idPart
    abstract = "abstract" ~idPart

    // Attributes
    nameAttribute = "@name"

    // Reserved
    reservedWord = keyword

    // Comments
    space += comment | lineTerminator
    comment = multiLineComment | singleLineComment
    lineTerminator = "\n" | "\r" | "\u2028" | "\u2029"
    multiLineComment = "/*" (~"*/" any)* "*/"
    singleLineComment = "//" (~lineTerminator any)*
}
