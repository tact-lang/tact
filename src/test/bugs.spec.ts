import { beginCell, toNano } from '@ton/core';
import { ContractSystem } from '@tact-lang/emulator';
import { __DANGER_resetNodeId } from '../grammar/ast';
import { SampleJetton } from './bugs/output/bugs_SampleJetton';
import { JettonDefaultWallet } from './bugs/output/bugs_JettonDefaultWallet';

describe('bugs', () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    it('should deploy contract correctly', async () => {

        // Init
        const system = await ContractSystem.create();
        const treasure = system.treasure('treasure');
        const contract = system.open(await SampleJetton.fromInit(treasure.address, beginCell().endCell(), toNano('100')));
        const target = system.open(await JettonDefaultWallet.fromInit(contract.address, treasure.address));
        const tracker = system.track(target.address);
        await contract.send(treasure, { value: toNano('10') }, { $$type: 'Mint', receiver: treasure.address, amount: toNano('10') });
        await system.run();

        expect(tracker.collect()).toMatchSnapshot();
    });
});