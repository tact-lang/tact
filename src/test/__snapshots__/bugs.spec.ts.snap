// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bugs should deploy contract correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{178D45190000000000000000502540BE400801D98D6D0FE85B55D6D58229C8ED55470B1C744D6BEEAD475C27236E9908A993110016E3A425A4E75B646191AC9A34FE5D050BD101A5C490F87D01C66D885D09BC1082_}",
            "type": "cell",
          },
          "bounce": false,
          "from": "kQDsxraH9C2q62rBFOR2qqOFjjomtfdWo64TkbdMhFTJiLsN",
          "to": "kQBGSDIgoUMAjGBNBXjsdBRlfPtqK5rKgQOD5N7yKFfIXeuh",
          "type": "internal",
          "value": "9.95885",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 16070n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "known",
              "value": {
                "$$type": "TokenExcesses",
                "queryId": 0n,
              },
            },
            "bounce": false,
            "from": "kQBGSDIgoUMAjGBNBXjsdBRlfPtqK5rKgQOD5N7yKFfIXeuh",
            "to": "@treasure(treasure)",
            "type": "internal",
            "value": "9.914852826",
          },
        ],
      },
    ],
  },
]
`;
